src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -249,10 +249,51 @@ private void SetUrl()
                 }
 
                 EventSourceTrace("Url: {0}", requestUri);
-                SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                if (requestUri.Host != requestUri.IdnHost)
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, ReplaceHostWithIdnHost(requestUri.AbsoluteUri, requestUri.Host, requestUri.IdnHost));
+                }
+                else
+                {
+                    SetCurlOption(CURLoption.CURLOPT_URL, requestUri.AbsoluteUri);
+                }
+
                 SetCurlOption(CURLoption.CURLOPT_PROTOCOLS, (long)(CurlProtocols.CURLPROTO_HTTP | CurlProtocols.CURLPROTO_HTTPS));
             }
 
+            private unsafe string ReplaceHostWithIdnHost(string absoluteUri, string host, string idnHost)
+            {
+                int start = absoluteUri.IndexOf(host);
```

Comment:
Can `requestUri` have a username or password component, and if so, what if one of those are equal to host (e.g. `"http://host:pass@host/path"`)? If that can happen, the code below will replace the first instance it finds, which would be the username/password instead of the host. I wonder if using `UriBuilder` to replace the host would work better (despite it likely allocating more). Something like (typing this on my phone; haven't actually tested it):

```c#
string host = requestUri.Host;
string idnHost = requestUri.IdnHost;
string absoluteUri = host == idnHost ?
    requestUri.AbsoluteUri :
    new UriBuilder(requestUri) { Host = idnHost }.Uri.AbsoluteUri;
SetCurlOption(CURLoption.CURLOPT_URL, absoluteUri);
```